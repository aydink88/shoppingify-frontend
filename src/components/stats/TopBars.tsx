import { useEffect, useMemo } from 'react';
import { ProgressBar, Stack } from 'react-bootstrap';

import useAsync from 'src/hooks/useAsync';

import TheSpinner from '../common/TheSpinner';

type TopBarsProps = {
  title: string;
  variant?: 'success' | 'danger' | 'warning' | 'info' | string;
  fetchData(): Promise<any>;
  howMany?: number;
};

export default function TopBars({ title, variant, fetchData, howMany = 3 }: TopBarsProps) {
  const { data, run } = useAsync<{ _id: string; name: string; amount: number }[]>([]);

  useEffect(() => {
    run(fetchData()).catch();
  }, []);

  const total = useMemo(() => {
    let t = 0;
    if (data?.length) {
      for (const item of data) {
        t += item.amount;
      }
    }
    return t;
  }, [data]);

  const dataSlice = useMemo(() => {
    if (!data?.length) return [];
    return data.slice(0, howMany);
  }, [data]);

  if (!data || !total) return <TheSpinner />;

  return (
    <div className="p-2 m-2">
      <h3>{title}</h3>
      {dataSlice.map((item) => {
        const percentage = (item.amount * 100) / total;
        return (
          <Stack key={item._id} className="my-3" data-testid="topbarsmember">
            <Stack direction="horizontal" className="justify-content-between">
              <span>{item.name}</span>
              <span>{percentage.toFixed(0)}%</span>
            </Stack>
            <ProgressBar
              variant={variant}
              now={percentage}
              label={`${percentage}%`}
              visuallyHidden
            />
          </Stack>
        );
      })}
    </div>
  );
}
