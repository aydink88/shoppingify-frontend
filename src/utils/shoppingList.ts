import type {
  TShoppingList,
  TShoppingListData,
  TShoppingListItem,
  TShoppingListWithoutItems,
} from 'src/types';

// prepare list data coming from backend
export function prepareActiveShoppingList(
  listsFromData: TShoppingListWithoutItems[]
): TShoppingList {
  let listForActiveListState: TShoppingListWithoutItems | undefined;

  for (const list of listsFromData) {
    if (!listForActiveListState && list.status === 'active') {
      listForActiveListState = list;
      break;
    }
  }

  // if there is an active list, items will be fetched later
  //const items={} as
  return listForActiveListState
    ? { ...listForActiveListState, items: {} }
    : {
        _id: '',
        name: 'Shopping List',
        items: {},
        status: 'active',
      };
}

// group list items by categories to fit into state
export function groupItemsOfShoppingList(list: TShoppingListData): TShoppingList {
  const newItems = {} as { [category: string]: TShoppingListItem[] };
  if (list.items?.length) {
    for (const item of list.items) {
      const category = item.item.category.name;
      if (category in newItems) {
        newItems[category].push(item);
      } else {
        newItems[category] = [item];
      }
    }
  }
  return {
    name: list.name,
    status: list.status,
    items: newItems,
    _id: list._id,
    createdAt: list.createdAt,
  };
}

export function mapShoppingListToSave(list: TShoppingList): TShoppingListData {
  //if (!Object.keys(list.items).length) return;
  const newItems: any = [];
  for (const categoryName in list.items) {
    list.items[categoryName].forEach((listItem) => {
      const newItem = { done: listItem.done, amount: listItem.amount, item: listItem.item._id };
      newItems.push(newItem);
    });
  }
  return { name: list.name, status: list.status, items: newItems, _id: list._id };
}

export function formatDate(date: number) {
  const formatter = new Intl.DateTimeFormat([], {
    weekday: 'short',
    day: 'numeric',
    month: 'numeric',
    year: 'numeric',
  });
  return formatter.format(new Date(date));
}
