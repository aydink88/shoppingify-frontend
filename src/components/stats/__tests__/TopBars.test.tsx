import { vi } from 'vitest';

import TopBars from 'src/components/stats/TopBars';
import { render, waitFor } from 'src/utils/test-utils';

describe('Top things chart using colored bars', () => {
  const mockItems = [
    { _id: 1, name: 'banana', amount: 3 },
    { _id: 2, name: 'apple', amount: 6 },
    { _id: 3, name: 'kiwi', amount: 4 },
    { _id: 4, name: 'cherry', amount: 1 },
  ];
  const fetchData = vi.fn().mockResolvedValue(mockItems);

  const renderComponent = () =>
    render(<TopBars fetchData={fetchData} title="Top Fruits" variant="success" />);

  it('should match snapshot', async () => {
    const { container } = await waitFor(renderComponent);
    expect(container).toMatchSnapshot();
  });

  it('should render data', async () => {
    const { findAllByTestId } = renderComponent();
    const items = await findAllByTestId('topbarsmember');
    expect(items.length).toBe(3);
  });
});
