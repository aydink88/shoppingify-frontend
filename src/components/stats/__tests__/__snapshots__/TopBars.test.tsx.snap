// Vitest Snapshot v1

exports[`Top things chart using colored bars should match snapshot 1`] = `
<div>
  <div
    class="p-2 m-2"
  >
    <h3>
      Top Fruits
    </h3>
    <div
      class="my-3 vstack"
      data-testid="topbarsmember"
    >
      <div
        class="justify-content-between hstack"
      >
        <span>
          banana
        </span>
        <span>
          21
          %
        </span>
      </div>
      <div
        class="progress"
      >
        <div
          aria-valuemax="100"
          aria-valuemin="0"
          aria-valuenow="21.428571428571427"
          class="progress-bar bg-success"
          role="progressbar"
          style="width: 21.429%;"
        >
          <span
            class="visually-hidden"
          >
            21.428571428571427%
          </span>
        </div>
      </div>
    </div>
    <div
      class="my-3 vstack"
      data-testid="topbarsmember"
    >
      <div
        class="justify-content-between hstack"
      >
        <span>
          apple
        </span>
        <span>
          43
          %
        </span>
      </div>
      <div
        class="progress"
      >
        <div
          aria-valuemax="100"
          aria-valuemin="0"
          aria-valuenow="42.857142857142854"
          class="progress-bar bg-success"
          role="progressbar"
          style="width: 42.857%;"
        >
          <span
            class="visually-hidden"
          >
            42.857142857142854%
          </span>
        </div>
      </div>
    </div>
    <div
      class="my-3 vstack"
      data-testid="topbarsmember"
    >
      <div
        class="justify-content-between hstack"
      >
        <span>
          kiwi
        </span>
        <span>
          29
          %
        </span>
      </div>
      <div
        class="progress"
      >
        <div
          aria-valuemax="100"
          aria-valuemin="0"
          aria-valuenow="28.571428571428573"
          class="progress-bar bg-success"
          role="progressbar"
          style="width: 28.571%;"
        >
          <span
            class="visually-hidden"
          >
            28.571428571428573%
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
